File: Writing-a-file/parallel_io.f90
! This is a parallel code where each process 
! writes its data to its own output file.

PROGRAM parallel_io
  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: wp=KIND(0.0D0)
  INTEGER, PARAMETER :: array_size=1000000
  INTEGER :: ierror
  CHARACTER(len=80) :: filename
  REAL(wp), DIMENSION(array_size) :: data1
  INTEGER :: rank, size
  REAL(wp) :: start, end

  CALL mpi_init(ierror)
  CALL mpi_comm_rank(MPI_COMM_WORLD,rank,ierror)
  CALL mpi_comm_size(MPI_COMM_WORLD,size,ierror)

! Initialise data1 array to different values on each process.
  data1 = REAL(rank,wp)

! Give each process a unique filename to write to.
  IF(rank<10)THEN
     WRITE(filename,'("test_out_file00",I1)') rank
  ELSE IF(rank<100) THEN
     WRITE(filename,'("test_out_file0",I2)') rank
  ELSE
     WRITE(filename,'("test_out_file",I3)') rank
  END IF
  OPEN(unit=17,file=filename,form='unformatted')

! Each process writes its data out to its own file.
  CALL mpi_barrier(MPI_COMM_WORLD,ierror)
  start = mpi_wtime()
  WRITE(17) data1
  CALL mpi_barrier(MPI_COMM_WORLD,ierror)
  end = mpi_wtime()

  CLOSE(17)
  IF(rank==0)THEN
     WRITE(6,'("Time taken was ",F10.6)') end-start
  END IF
  CALL mpi_finalize(ierror)

END PROGRAM parallel_io

File: Writing-a-file/parallel_io.c
/* This is a parallel code where each process 
   writes its data to its own output file. */

#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char *argv[])
{
  const int array_size=1000000;
  int rank, size, i;
  char filesuffix[80], filename[80];
  double data1[array_size];
  double start, end;
  FILE *fh;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

/* Initialise data1 array to different values on each process. */
  for (i=0; i<array_size; i++)
     data1[i] = (double)rank;

/* Give each process a unique filename to write to. */
  if(rank < 10){
     strcpy(filesuffix,"test_out_file00");
  }
  else if(rank<100){
     strcpy(filesuffix,"test_out_file0");
  }
  else{
     strcpy(filesuffix,"test_out_file");
  }
  sprintf(filename, "%s%d", filesuffix, rank);
  fh=fopen(filename,"wb");
  
/* Each process writes its data out to its own file. */
  MPI_Barrier(MPI_COMM_WORLD);
  start = MPI_Wtime();
  fwrite(data1,sizeof(double),array_size,fh);
  MPI_Barrier(MPI_COMM_WORLD);
  end = MPI_Wtime();

  fclose(fh);
  if(rank==0){
     printf("Time taken was %g seconds.\n", end-start);
  }
  MPI_Finalize();
  return 0;
}
